### PR TITLE
Super agent health reporting

### PR DESCRIPTION
**Pull request by jtduffy**
Implementation of file based health check reporting to the super agent/NR control.

In-flight agent spec: https://source.datanerd.us/agents/agent-specs/pull/690/files

### CODE REVIEW

File: newrelic-agent/src/main/java/com/newrelic/agent/superagent/SuperAgentIntegrationService.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        for (HealthDataProducer healthProducer : healthProducers) {
+            healthProducer.registerHealthDataChangeListener(this);
+        }
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        if (isEnabled()) {
+            Agent.LOG.log(Level.INFO, "SuperAgentIntegrationService starting: Health file location: {0}  Frequency: {1}   Scheme: {2}",
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/superagent/SuperAgentIntegrationService.java:48]:
```suggestion
            Agent.LOG.log(Level.INFO, "SuperAgentIntegrationService starting: Health file location: {0}  Frequency: {1}  Scheme: {2}",
```

File: newrelic-agent/src/test/java/com/newrelic/agent/AgentTest.java [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import com.newrelic.agent.config.AgentConfig;
```
Comment (jasonjkeller) [inline newrelic-agent/src/test/java/com/newrelic/agent/AgentTest.java:3]: Needs a copyright header

File: newrelic-agent/src/test/java/com/newrelic/agent/superagent/SuperAgentHealthUnitTestClient.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/*
+ *
+ *  * Copyright 2024 New Relic Corporation. All rights reserved.
+ *  * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package com.newrelic.agent.superagent;
+
+public class
+SuperAgentHealthUnitTestClient implements SuperAgentIntegrationHealthClient {
```
Comment (jasonjkeller) [inline newrelic-agent/src/test/java/com/newrelic/agent/superagent/SuperAgentHealthUnitTestClient.java:10]:
```suggestion
public class SuperAgentHealthUnitTestClient implements SuperAgentIntegrationHealthClient {
```

File: newrelic-agent/src/main/java/com/newrelic/agent/superagent/SuperAgentIntegrationUtils.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        }
+    }
+
+    public static void reportHealthyStatus(List<HealthDataChangeListener> healthDataChangeListeners, AgentHealth.Category... categories) {
+        for (HealthDataChangeListener listener : healthDataChangeListeners) {
+            listener.onHealthyStatus(categories);
+        }
+    }
+
+    public static void  reportUnhealthyStatusPriorToServiceStart(AgentConfig config, AgentHealth.Status status) {
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/superagent/SuperAgentIntegrationUtils.java:34]:
```suggestion
    public static void reportUnhealthyStatusPriorToServiceStart(AgentConfig config, AgentHealth.Status status) {
```

File: newrelic-agent/src/main/java/com/newrelic/agent/transport/DataSenderImpl.java [hunk 5]
```diff
@@ -669,15 +680,19 @@ private void throwExceptionFromStatusCode(String method, ReadResult result, byte
                 }
             case HttpResponseCode.UNAUTHORIZED:
                 // received 401 Unauthorized, throw exception instead of parsing LicenseException from 200 response body
+                SuperAgentIntegrationUtils.reportUnhealthyStatus(healthDataChangeListeners, AgentHealth.Status.INVALID_LICENSE);
```
Comment (jasonjkeller) [inline newrelic-agent/src/main/java/com/newrelic/agent/transport/DataSenderImpl.java:683]: Any concerns about having this stuff mixed into the Datasender if the super agent is disabled?

**Review by jasonjkeller — CHANGES_REQUESTED** [review]
Looks good overall. Just a couple of questions and minor cleanup suggestions.

File: newrelic-agent/src/main/java/com/newrelic/agent/transport/DataSenderImpl.java (hunk 5 above)
Comment (jtduffy) [reply newrelic-agent/src/main/java/com/newrelic/agent/transport/DataSenderImpl.java:683]: Good idea. I'll refactor and wrap these calls around a config check gate just to be double safe

**Review by jasonjkeller — APPROVED** [review]

**Pull request merged**